# Housing Rental Analysis for San Francisco

In this challenge, your job is to use your data visualization skills, including aggregation, interactive visualizations, and geospatial analysis, to find properties in the San Francisco market that are viable investment opportunities.

## Instructions

Use the `san_francisco_housing.ipynb` notebook to visualize and analyze the real-estate data.

Note that this assignment requires you to create a visualization by using hvPlot and GeoViews. Additionally, you need to read the `sfo_neighborhoods_census_data.csv` file from the `Resources` folder into the notebook and create the DataFrame that you’ll use in the analysis.

The main task in this Challenge is to visualize and analyze the real-estate data in your Jupyter notebook. Use the `san_francisco_housing.ipynb` notebook to complete the following tasks:

* Calculate and plot the housing units per year.

* Calculate and plot the average prices per square foot.

* Compare the average prices by neighborhood.

* Build an interactive neighborhood map.

* Compose your data story.

### Calculate and Plot the Housing Units per Year

For this part of the assignment, use numerical and visual aggregation to calculate the number of housing units per year, and then visualize the results as a bar chart. To do so, complete the following steps:

1. Use the `groupby` function to group the data by year. Aggregate the results by the `mean` of the groups.

2. Use the `hvplot` function to plot the `housing_units_by_year` DataFrame as a bar chart. Make the x-axis represent the `year` and the y-axis represent the `housing_units`.

3. Style and format the line plot to ensure a professionally styled visualization.

4. Note that your resulting plot should appear similar to the following image:

![A screenshot depicts an example of the resulting bar chart.](Images/zoomed-housing-units-by-year.png)

5. Answer the following question:

    * What’s the overall trend in housing units over the period that you’re analyzing?

### Calculate and Plot the Average Sale Prices per Square Foot

For this part of the assignment, use numerical and visual aggregation to calculate the average prices per square foot, and then visualize the results as a bar chart. To do so, complete the following steps:

1. Group the data by year, and then average the results. What’s the lowest gross rent that’s reported for the years that the DataFrame includes?

2. Create a new DataFrame named `prices_square_foot_by_year` by filtering out the “housing_units” column. The new DataFrame should include the averages per year for only the sale price per square foot and the gross rent.

3. Use hvPlot to plot the `prices_square_foot_by_year` DataFrame as a line plot.

    > **Hint** This single plot will include lines for both `sale_price_sqr_foot` and `gross_rent`.

4. Style and format the line plot to ensure a professionally styled visualization.

5. Note that your resulting plot should appear similar to the following image:

![A screenshot depicts an example of the resulting plot.](Images/avg-sale-px-sq-foot-gross-rent.png)

6. Use both the `prices_square_foot_by_year` DataFrame and interactive plots to answer the following questions:

    * Did any year experience a drop in the average sale price per square foot compared to the previous year?

    * If so, did the gross rent increase or decrease during that year?

### Compare the Average Sale Prices by Neighborhood

For this part of the assignment, use interactive visualizations and widgets to explore the average sale price per square foot by neighborhood. To do so, complete the following steps:

1. Create a new DataFrame that groups the original DataFrame by year and neighborhood. Aggregate the results by the `mean` of the groups.

2. Filter out the “housing_units” column to create a DataFrame that includes only the `sale_price_sqr_foot` and `gross_rent` averages per year.

3. Create an interactive line plot with hvPlot that visualizes both `sale_price_sqr_foot` and `gross_rent`. Set the x-axis parameter to the year (`x="year"`). Use the `groupby` parameter to create an interactive widget for `neighborhood`.

4. Style and format the line plot to ensure a professionally styled visualization.

5. Note that your resulting plot should appear similar to the following image:

![A screenshot depicts an example of the resulting plot.](Images/pricing-info-by-neighborhood.png)

6. Use the interactive visualization to answer the following question:

    * For the Anza Vista neighborhood, is the average sale price per square foot for 2016 more or less than the price that’s listed for 2012? 

### Build an Interactive Neighborhood Map

For this part of the assignment, explore the geospatial relationships in the data by using interactive visualizations with hvPlot and GeoViews. To build your map, use the `sfo_data_df` DataFrame (created during the initial import), which includes the neighborhood location data with the average prices. To do all this, complete the following steps:

1. Read the `neighborhood_coordinates.csv` file from the `Resources` folder into the notebook, and create a DataFrame named `neighborhood_locations_df`. Be sure to set the `index_col` of the DataFrame as “Neighborhood”.

2. Using the original `sfo_data_df` Dataframe, create a DataFrame named `all_neighborhood_info_df` that groups the data by neighborhood. Aggregate the results by the `mean` of the group.

3. Review the two code cells that concatenate the `neighborhood_locations_df` DataFrame with the `all_neighborhood_info_df` DataFrame. Note that the first cell uses the [Pandas concat function](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.concat.html) to create a DataFrame named `all_neighborhoods_df`. The second cell cleans the data and sets the “Neighborhood” column. Be sure to run these cells to create the `all_neighborhoods_df` DataFrame, which you’ll need to create the geospatial visualization.

4. Using hvPlot with GeoViews enabled, create a `points` plot for the `all_neighborhoods_df` DataFrame. Be sure to do the following:

    * Set the `geo` parameter to True.
    * Set the `size` parameter to “sale_price_sqr_foot”.
    * Set the `color` parameter to “gross_rent”.
    * Set the `frame_width` parameter to 700.
    * Set the `frame_height` parameter to 500.
    * Include a descriptive title.

Note that your resulting plot should appear similar to the following image:

![A screenshot depicts an example of a scatter plot created with hvPlot and GeoViews.](Images/6-4-geoviews-plot.png)

5. Use the interactive map to answer the following question:

    * Which neighborhood has the highest gross rent, and which has the highest sale price per square foot?

### Compose Your Data Story

Based on the visualizations that you created, answer the following questions:

* How does the trend in rental income growth compare to the trend in sales prices? Does this same trend hold true for all the neighborhoods across San Francisco?

* What insights can you share with your company about the potential one-click, buy-and-rent strategy that they're pursuing? Do neighborhoods exist that you would suggest for investment, and why?

In [47]:
# Import the required libraries and dependencies
import pandas as pd
import hvplot.pandas
from pathlib import Path
#Import libriairies needed for project

## Import the data 

In [48]:
# Using the read_csv function and Path module, create a DataFrame 
# by importing the sfo_neighborhoods_census_data.csv file from the Resources folder
sfo_data_df = pd.read_csv(
    Path("./Resources/sfo_neighborhoods_census_data.csv")
)
#Create a dataframe using the read_csv function
# Review the first and last five rows of the DataFrame
sfo_data_df.head()
#review first five rows of data frame.

year      neighborhood  sale_price_sqr_foot  housing_units  gross_rent
0  2010      Alamo Square           291.182945         372560        1239
1  2010        Anza Vista           267.932583         372560        1239
2  2010           Bayview           170.098665         372560        1239
3  2010  Buena Vista Park           347.394919         372560        1239
4  2010  Central Richmond           319.027623         372560        1239

In [49]:
sfo_data_df.tail()
#Review last five rows of data frame.

year            neighborhood  sale_price_sqr_foot  housing_units  \
392  2016          Telegraph Hill           903.049771         384242   
393  2016              Twin Peaks           970.085470         384242   
394  2016  Van Ness/ Civic Center           552.602567         384242   
395  2016       Visitacion Valley           328.319007         384242   
396  2016           Westwood Park           631.195426         384242   

     gross_rent  
392        4390  
393        4390  
394        4390  
395        4390  
396        4390

---

## Calculate and Plot the Housing Units per Year

For this part of the assignment, use numerical and visual aggregation to calculate the number of housing units per year, and then visualize the results as a bar chart. To do so, complete the following steps:

1. Use the `groupby` function to group the data by year. Aggregate the results by the `mean` of the groups.

2. Use the `hvplot` function to plot the `housing_units_by_year` DataFrame as a bar chart. Make the x-axis represent the `year` and the y-axis represent the `housing_units`.

3. Style and format the line plot to ensure a professionally styled visualization.

4. Note that your resulting plot should appear similar to the following image:

![A screenshot depicts an example of the resulting bar chart.](Images/zoomed-housing-units-by-year.png)

5. Answer the following question:

    * What’s the overall trend in housing units over the period that you’re analyzing?



### Step 1: Use the `groupby` function to group the data by year. Aggregate the results by the `mean` of the groups.

In [50]:
# Create a numerical aggregation that groups the data by the year and then averages the results.
housing_units_by_year = sfo_data_df.groupby("year")["housing_units"].mean()

# Review the DataFrame
housing_units_by_year
#Using groupby function gather the mean of housing units by year and display the average results.

year
2010    372560.0
2011    374507.0
2012    376454.0
2013    378401.0
2014    380348.0
2015    382295.0
2016    384242.0
Name: housing_units, dtype: float64

### Step 2: Use the `hvplot` function to plot the `housing_units_by_year` DataFrame as a bar chart. Make the x-axis represent the `year` and the y-axis represent the `housing_units`.

### Step 3: Style and format the line plot to ensure a professionally styled visualization.

In [51]:
# Create a visual aggregation explore the housing units by year
housing_units_by_year.hvplot.bar( x='year', y='housing_units', title="Housing Units in San Francisco from 2010 to 2016", xlabel="Year", ylabel="Housing Units", ylim=(365000,385000) ).opts(yformatter='%.0f')
#Create a bar chart that displays the year on the x axis and housing units on the y axis and label them accordingly.  Also, add the title along with the format for housing units and start the housing units at 365,000 instead of 0 to show the grouth.


:Bars   [year]   (housing_units)

### Step 5: Answer the following question:

**Question:** What is the overall trend in housing_units over the period being analyzed?

**Answer:** The overall trend of housing units in San Francisco consistently increases from 2010 with a mean of 372,560 to a mean of 384,242 in 2016.

---

## Calculate and Plot the Average Sale Prices per Square Foot

For this part of the assignment, use numerical and visual aggregation to calculate the average prices per square foot, and then visualize the results as a bar chart. To do so, complete the following steps:

1. Group the data by year, and then average the results. What’s the lowest gross rent that’s reported for the years that the DataFrame includes?

2. Create a new DataFrame named `prices_square_foot_by_year` by filtering out the “housing_units” column. The new DataFrame should include the averages per year for only the sale price per square foot and the gross rent.

3. Use hvPlot to plot the `prices_square_foot_by_year` DataFrame as a line plot.

    > **Hint** This single plot will include lines for both `sale_price_sqr_foot` and `gross_rent`.

4. Style and format the line plot to ensure a professionally styled visualization.

5. Note that your resulting plot should appear similar to the following image:

![A screenshot depicts an example of the resulting plot.](Images/avg-sale-px-sq-foot-gross-rent.png)

6. Use both the `prices_square_foot_by_year` DataFrame and interactive plots to answer the following questions:

    * Did any year experience a drop in the average sale price per square foot compared to the previous year?

    * If so, did the gross rent increase or decrease during that year?



### Step 1: Group the data by year, and then average the results.

In [52]:
# Create a numerical aggregation by grouping the data by year and averaging the results
prices_square_foot_by_year = sfo_data_df.groupby("year").mean()

# Review the resulting DataFrame
prices_square_foot_by_year
#Group the data frame by year and show the results.


sale_price_sqr_foot  housing_units  gross_rent
year                                                
2010           369.344353       372560.0      1239.0
2011           341.903429       374507.0      1530.0
2012           399.389968       376454.0      2324.0
2013           483.600304       378401.0      2971.0
2014           556.277273       380348.0      3528.0
2015           632.540352       382295.0      3739.0
2016           697.643709       384242.0      4390.0

**Question:** What is the lowest gross rent reported for the years included in the DataFrame?

**Answer:** 2010 has the lowest gross rent reported with $1,239.00.



### Step 2: Create a new DataFrame named `prices_square_foot_by_year` by filtering out the “housing_units” column. The new DataFrame should include the averages per year for only the sale price per square foot and the gross rent.

In [53]:
# Filter out the housing_units column, creating a new DataFrame 
# Keep only sale_price_sqr_foot and gross_rent averages per year
prices_square_foot_by_year = prices_square_foot_by_year.drop(columns=["housing_units"])

# Review the DataFrame
prices_square_foot_by_year
#Remove column labeled housing units and display results.

sale_price_sqr_foot  gross_rent
year                                 
2010           369.344353      1239.0
2011           341.903429      1530.0
2012           399.389968      2324.0
2013           483.600304      2971.0
2014           556.277273      3528.0
2015           632.540352      3739.0
2016           697.643709      4390.0

### Step 3: Use hvPlot to plot the `prices_square_foot_by_year` DataFrame as a line plot.

> **Hint** This single plot will include lines for both `sale_price_sqr_foot` and `gross_rent`

### Step 4: Style and format the line plot to ensure a professionally styled visualization.


In [54]:
# Plot prices_square_foot_by_year. 
# Inclued labels for the x- and y-axes, and a title.
prices_square_foot_by_year.hvplot.line(x="year",yaxis=True, xlabel="Year", ylabel="Gross Rent / Sale Price per Sq Foot", title= "Sales Price per Sq Foot and Average Gross Rent 2010 - 2016 San Francisco.", size=(35,7))
#plot a line chart with average gross rent and sales price per square foot for the year 2010 to 2016 in San Francisco.  Using the x axis as the year and y axis as the avg gross rent and price per sq foot and labeling them as such.  Then adding a title and sizing the chart.

:NdOverlay   [Variable]
   :Curve   [year]   (value)

### Step 6: Use both the `prices_square_foot_by_year` DataFrame and interactive plots to answer the following questions:

**Question:** Did any year experience a drop in the average sale price per square foot compared to the previous year?

**Answer:** #Yes, price per square foot decreases from 369.34 in the year 2010 to 341.90 in 2011.

**Question:** If so, did the gross rent increase or decrease during that year?

**Answer:** Eventhough the price per square foot decreased in 2011 average gross rent increased from 1,239 to 1,530 in the same year.

---

## Compare the Average Sale Prices by Neighborhood

For this part of the assignment, use interactive visualizations and widgets to explore the average sale price per square foot by neighborhood. To do so, complete the following steps:

1. Create a new DataFrame that groups the original DataFrame by year and neighborhood. Aggregate the results by the `mean` of the groups.

2. Filter out the “housing_units” column to create a DataFrame that includes only the `sale_price_sqr_foot` and `gross_rent` averages per year.

3. Create an interactive line plot with hvPlot that visualizes both `sale_price_sqr_foot` and `gross_rent`. Set the x-axis parameter to the year (`x="year"`). Use the `groupby` parameter to create an interactive widget for `neighborhood`.

4. Style and format the line plot to ensure a professionally styled visualization.

5. Note that your resulting plot should appear similar to the following image:

![A screenshot depicts an example of the resulting plot.](Images/pricing-info-by-neighborhood.png)

6. Use the interactive visualization to answer the following question:

    * For the Anza Vista neighborhood, is the average sale price per square foot for 2016 more or less than the price that’s listed for 2012? 


### Step 1: Create a new DataFrame that groups the original DataFrame by year and neighborhood. Aggregate the results by the `mean` of the groups.

In [55]:
# Group by year and neighborhood and then create a new dataframe of the mean values
prices_by_year_by_neighborhood = sfo_data_df.groupby(["year","neighborhood"]).mean()
# Review the DataFrame
prices_by_year_by_neighborhood
#Group year and neighborhood to get average (mean) results.

sale_price_sqr_foot  housing_units  gross_rent
year neighborhood                                                          
2010 Alamo Square                     291.182945       372560.0      1239.0
     Anza Vista                       267.932583       372560.0      1239.0
     Bayview                          170.098665       372560.0      1239.0
     Buena Vista Park                 347.394919       372560.0      1239.0
     Central Richmond                 319.027623       372560.0      1239.0
...                                          ...            ...         ...
2016 Telegraph Hill                   903.049771       384242.0      4390.0
     Twin Peaks                       970.085470       384242.0      4390.0
     Van Ness/ Civic Center           552.602567       384242.0      4390.0
     Visitacion Valley                328.319007       384242.0      4390.0
     Westwood Park                    631.195426       384242.0      4390.0

[397 rows x 3 columns]

### Step 2: Filter out the “housing_units” column to create a DataFrame that includes only the `sale_price_sqr_foot` and `gross_rent` averages per year.

In [56]:
# Filter out the housing_units
prices_by_year_by_neighborhood = prices_by_year_by_neighborhood.drop(columns=["housing_units"])


# Review the first and last five rows of the DataFrame
prices_by_year_by_neighborhood.head()
# Remove Housing units columns and display top 5 rows and bottom 5 rows.

sale_price_sqr_foot  gross_rent
year neighborhood                                     
2010 Alamo Square               291.182945      1239.0
     Anza Vista                 267.932583      1239.0
     Bayview                    170.098665      1239.0
     Buena Vista Park           347.394919      1239.0
     Central Richmond           319.027623      1239.0

In [57]:
prices_by_year_by_neighborhood.tail()

sale_price_sqr_foot  gross_rent
year neighborhood                                           
2016 Telegraph Hill                   903.049771      4390.0
     Twin Peaks                       970.085470      4390.0
     Van Ness/ Civic Center           552.602567      4390.0
     Visitacion Valley                328.319007      4390.0
     Westwood Park                    631.195426      4390.0

### Step 3: Create an interactive line plot with hvPlot that visualizes both `sale_price_sqr_foot` and `gross_rent`. Set the x-axis parameter to the year (`x="year"`). Use the `groupby` parameter to create an interactive widget for `neighborhood`.

### Step 4: Style and format the line plot to ensure a professionally styled visualization.

In [58]:
# Use hvplot to create an interactive line plot of the average price per square foot
# The plot should have a dropdown selector for the neighborhood
prices_by_year_by_neighborhood.hvplot(
    groupby="neighborhood", 
    x="year",
    rot=90, 
    size=(35, 7),
    xlabel="Year",
    ylabel="Gross Rent / Sales Price per Square Foot",
    title="Sales Price per Square Foot and Average Gross Rent - 2010-2016 By Neighborhood"
)
#Create a line plot with an interactive widget for neighborhood.  Label the x and y axis along with creating a title for the line plot.

:DynamicMap   [neighborhood]
   :NdOverlay   [Variable]
      :Curve   [year]   (value)

### Step 6: Use the interactive visualization to answer the following question:

**Question:** For the Anza Vista neighborhood, is the average sale price per square foot for 2016 more or less than the price that’s listed for 2012? 

**Answer:** The average sale price per square foot for 2016 in the Anza Visata neighborhood is less in 2016 at 88.40 compared to 2010 at 267.93.

---

## Build an Interactive Neighborhood Map

For this part of the assignment, explore the geospatial relationships in the data by using interactive visualizations with hvPlot and GeoViews. To build your map, use the `sfo_data_df` DataFrame (created during the initial import), which includes the neighborhood location data with the average prices. To do all this, complete the following steps:

1. Read the `neighborhood_coordinates.csv` file from the `Resources` folder into the notebook, and create a DataFrame named `neighborhood_locations_df`. Be sure to set the `index_col` of the DataFrame as “Neighborhood”.

2. Using the original `sfo_data_df` Dataframe, create a DataFrame named `all_neighborhood_info_df` that groups the data by neighborhood. Aggregate the results by the `mean` of the group.

3. Review the two code cells that concatenate the `neighborhood_locations_df` DataFrame with the `all_neighborhood_info_df` DataFrame. Note that the first cell uses the [Pandas concat function](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.concat.html) to create a DataFrame named `all_neighborhoods_df`. The second cell cleans the data and sets the “Neighborhood” column. Be sure to run these cells to create the `all_neighborhoods_df` DataFrame, which you’ll need to create the geospatial visualization.

4. Using hvPlot with GeoViews enabled, create a `points` plot for the `all_neighborhoods_df` DataFrame. Be sure to do the following:

    * Set the `size` parameter to “sale_price_sqr_foot”.

    * Set the `color` parameter to “gross_rent”.

    * Set the `size_max` parameter to “25”.

    * Set the `zoom` parameter to “11”.

Note that your resulting plot should appear similar to the following image:

![A screenshot depicts an example of a scatter plot created with hvPlot and GeoViews.](Images/6-4-geoviews-plot.png)

5. Use the interactive map to answer the following question:

    * Which neighborhood has the highest gross rent, and which has the highest sale price per square foot?


### Step 1: Read the `neighborhood_coordinates.csv` file from the `Resources` folder into the notebook, and create a DataFrame named `neighborhood_locations_df`. Be sure to set the `index_col` of the DataFrame as “Neighborhood”.

In [59]:
# Load neighborhoods coordinates data
neighborhood_locations_df = pd.read_csv(
    Path("./Resources/neighborhoods_coordinates.csv"),
    index_col= "Neighborhood"
)

# Review the DataFrame
neighborhood_locations_df.head()
#Using pandas read the csv file and review the dataframe.

Lat         Lon
Neighborhood                          
Alamo Square     37.791012 -122.402100
Anza Vista       37.779598 -122.443451
Bayview          37.734670 -122.401060
Bayview Heights  37.728740 -122.410980
Bernal Heights   37.728630 -122.443050

### Step 2: Using the original `sfo_data_df` Dataframe, create a DataFrame named `all_neighborhood_info_df` that groups the data by neighborhood. Aggregate the results by the `mean` of the group.

In [60]:
# Calculate the mean values for each neighborhood
all_neighborhood_info_df = sfo_data_df.groupby(["neighborhood"]).mean().round()

# Review the resulting DataFrame
all_neighborhood_info_df
#Using original sfo_data_df dataframe group neighborhoods and average (mean) the results.

year  sale_price_sqr_foot  housing_units  gross_rent
neighborhood                                                              
Alamo Square        2013.0                366.0       378401.0      2817.0
Anza Vista          2013.0                373.0       379050.0      3032.0
Bayview             2012.0                205.0       376454.0      2318.0
Bayview Heights     2015.0                591.0       382295.0      3739.0
Bernal Heights      2014.0                577.0       379374.0      3080.0
...                    ...                  ...            ...         ...
West Portal         2012.0                498.0       376941.0      2516.0
Western Addition    2012.0                308.0       377428.0      2555.0
Westwood Highlands  2012.0                534.0       376454.0      2250.0
Westwood Park       2015.0                687.0       382295.0      3959.0
Yerba Buena         2012.0                577.0       377428.0      2555.0

[73 rows x 4 columns]

### Step 3: Review the two code cells that concatenate the `neighborhood_locations_df` DataFrame with the `all_neighborhood_info_df` DataFrame. 

Note that the first cell uses the [Pandas concat function](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.concat.html) to create a DataFrame named `all_neighborhoods_df`. 

The second cell cleans the data and sets the “Neighborhood” column. 

Be sure to run these cells to create the `all_neighborhoods_df` DataFrame, which you’ll need to create the geospatial visualization.

In [61]:
# Using the Pandas `concat` function, join the 
# neighborhood_locations_df and the all_neighborhood_info_df DataFrame
# The axis of the concatenation is "columns".
# The concat function will automatially combine columns with
# identical information, while keeping the additional columns.
all_neighborhoods_df = pd.concat(
    [neighborhood_locations_df, all_neighborhood_info_df], 
    axis="columns",
    sort=False
)

# Review the resulting DataFrame
display(all_neighborhoods_df)
#Joins the all__neighborhood_info_df and neighborhood_locations_df dataframes by using the concat function.

Lat         Lon    year  sale_price_sqr_foot  \
Alamo Square     37.791012 -122.402100  2013.0                366.0   
Anza Vista       37.779598 -122.443451  2013.0                373.0   
Bayview          37.734670 -122.401060  2012.0                205.0   
Bayview Heights  37.728740 -122.410980  2015.0                591.0   
Bernal Heights   37.728630 -122.443050     NaN                  NaN   
...                    ...         ...     ...                  ...   
Yerba Buena      37.792980 -122.396360  2012.0                577.0   
Bernal Heights         NaN         NaN  2014.0                577.0   
Downtown               NaN         NaN  2013.0                391.0   
Ingleside              NaN         NaN  2012.0                368.0   
Outer Richmond         NaN         NaN  2013.0                474.0   

                 housing_units  gross_rent  
Alamo Square          378401.0      2817.0  
Anza Vista            379050.0      3032.0  
Bayview               376454.0      2318.0  
Bayview Heights       382295.0      3739.0  
Bernal Heights             NaN         NaN  
...                        ...         ...  
Yerba Buena           377428.0      2555.0  
Bernal Heights        379374.0      3080.0  
Downtown              378401.0      2817.0  
Ingleside             377428.0      2509.0  
Outer Richmond        378401.0      2817.0  

[77 rows x 6 columns]

In [65]:
all_neighborhoods_df.sort_values("gross_rent")
                                 

Neighborhood       Lat         Lon    year  sale_price_sqr_foot  \
19       Forest Knolls  37.75706 -122.455090  2011.0                322.0   
12     Diamond Heights  37.72863 -122.443050  2012.0                435.0   
60         South Beach  37.78323 -122.400650  2012.0                650.0   
38       Miraloma Park  37.73415 -122.457000  2012.0                780.0   
70  Westwood Highlands  37.73470 -122.456854  2012.0                534.0   
..                 ...       ...         ...     ...                  ...   
36      Merced Heights  37.71993 -122.465950  2014.0                789.0   
59      Silver Terrace  37.73467 -122.401060  2014.0                170.0   
67   Visitacion Valley  37.72874 -122.410980  2014.0                301.0   
3      Bayview Heights  37.72874 -122.410980  2015.0                591.0   
71       Westwood Park  37.73415 -122.457000  2015.0                687.0   

    housing_units  gross_rent  
19       374507.0      1782.0  
12       375480.0      2016.0  
60       375805.0      2099.0  
38       375967.0      2155.0  
70       376454.0      2250.0  
..            ...         ...  
36       380348.0      3414.0  
59       380348.0      3528.0  
67       381322.0      3657.0  
3        382295.0      3739.0  
71       382295.0      3959.0  

[69 rows x 7 columns]

In [66]:
all_neighborhoods_df.sort_values("sale_price_sqr_foot")

Neighborhood       Lat        Lon    year  sale_price_sqr_foot  \
59         Silver Terrace  37.73467 -122.40106  2014.0                170.0   
24          Hunters Point  37.72551 -122.37178  2012.0                171.0   
2                 Bayview  37.73467 -122.40106  2012.0                205.0   
47          Outer Mission  37.72280 -122.43869  2014.0                242.0   
67      Visitacion Valley  37.72874 -122.41098  2014.0                301.0   
..                    ...       ...        ...     ...                  ...   
71          Westwood Park  37.73415 -122.45700  2015.0                687.0   
51        Pacific Heights  37.79298 -122.43579  2013.0                690.0   
38          Miraloma Park  37.73415 -122.45700  2012.0                780.0   
36         Merced Heights  37.71993 -122.46595  2014.0                789.0   
65  Union Square District  37.79101 -122.40210  2012.0                904.0   

    housing_units  gross_rent  
59       380348.0      3528.0  
24       377428.0      2489.0  
2        376454.0      2318.0  
47       379374.0      2996.0  
67       381322.0      3657.0  
..            ...         ...  
71       382295.0      3959.0  
51       378401.0      2817.0  
38       375967.0      2155.0  
36       380348.0      3414.0  
65       377428.0      2555.0  

[69 rows x 7 columns]

In [62]:
# Call the dropna function to remove any neighborhoods that do not have data
all_neighborhoods_df = all_neighborhoods_df.reset_index().dropna()

# Rename the "index" column as "Neighborhood" for use in the Visualization
all_neighborhoods_df = all_neighborhoods_df.rename(columns={"index": "Neighborhood"})


# Review the resulting DataFrame
display(all_neighborhoods_df.head())
display(all_neighborhoods_df.tail())

Neighborhood        Lat         Lon    year  sale_price_sqr_foot  \
0      Alamo Square  37.791012 -122.402100  2013.0                366.0   
1        Anza Vista  37.779598 -122.443451  2013.0                373.0   
2           Bayview  37.734670 -122.401060  2012.0                205.0   
3   Bayview Heights  37.728740 -122.410980  2015.0                591.0   
5  Buena Vista Park  37.768160 -122.439330  2013.0                453.0   

   housing_units  gross_rent  
0       378401.0      2817.0  
1       379050.0      3032.0  
2       376454.0      2318.0  
3       382295.0      3739.0  
5       378076.0      2699.0

Neighborhood       Lat         Lon    year  sale_price_sqr_foot  \
68         West Portal  37.74026 -122.463880  2012.0                498.0   
69    Western Addition  37.79298 -122.435790  2012.0                308.0   
70  Westwood Highlands  37.73470 -122.456854  2012.0                534.0   
71       Westwood Park  37.73415 -122.457000  2015.0                687.0   
72         Yerba Buena  37.79298 -122.396360  2012.0                577.0   

    housing_units  gross_rent  
68       376941.0      2516.0  
69       377428.0      2555.0  
70       376454.0      2250.0  
71       382295.0      3959.0  
72       377428.0      2555.0

### Step 4: Using hvPlot with GeoViews enabled, create a `points` plot for the `all_neighborhoods_df` DataFrame. Be sure to do the following:

* Set the `geo` parameter to True.
* Set the `size` parameter to “sale_price_sqr_foot”.
* Set the `color` parameter to “gross_rent”.
* Set the `frame_width` parameter to 700.
* Set the `frame_height` parameter to 500.
* Include a descriptive title.

In [63]:
# Create a plot to analyze neighborhood info
all_neighborhoods_df.hvplot.points(
    "Lon",
    "Lat",
    geo=True,
    tiles="OSM",
    size="sale_price_sqr_foot",
    color="gross_rent",
    frame_width=700,
    frame_height=500,
    title="Average Sale Price Per Square Foot and Gross Rent in San Francisco"
)



    

:Overlay
   .Tiles.I  :Tiles   [x,y]
   .Points.I :Points   [Lon,Lat]   (gross_rent,sale_price_sqr_foot)

### Step 5: Use the interactive map to answer the following question:

**Question:** Which neighborhood has the highest gross rent, and which has the highest sale price per square foot?

**Answer:** # Westwood park has the highest gross rent and Union Square has the highest sale price per square foot.


## Compose Your Data Story

Based on the visualizations that you have created, compose a data story that synthesizes your analysis by answering the following questions:

**Question:**  How does the trend in rental income growth compare to the trend in sales prices? Does this same trend hold true for all the neighborhoods across San Francisco?

**Answer:** The rental income increases based on sales price per square foot however it does not hold the same trend for each neighborhood.  The deeper the blue the better the gross rent per square foot based on the neighborhood is shown on the map.

**Question:** What insights can you share with your company about the potential one-click, buy-and-rent strategy that they're pursuing? Do neighborhoods exist that you would suggest for investment, and why?

**Answer:** Look for low price per square foot amounts compared to gross rent.  One that stands out is Silver Terrace at 170 price per square foot with a gross rent of $3528.  The low end of price per square foot and the high end of gross rent.